solver/jobs_test.go

```diff
@@ -0,0 +1,73 @@
+package solver
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/client/llb"
+	"github.com/moby/buildkit/util/testutil/integration"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+)
+
+func init() {
+	integration.InitOCIWorker()
+}
+
+func TestJobsIntegration(t *testing.T) {
+	mirrors := integration.WithMirroredImages(integration.OfficialImages("busybox:latest"))
+
+	integration.Run(t, []integration.Test{
+		testParallelism,
+	},
+		mirrors,
+		integration.WithMatrix("max-parallelism", map[string]interface{}{
+			"single": maxParallelismSingle,
+		}),
+	)
+}
+
+func testParallelism(t *testing.T, sb integration.Sandbox) {
+	ctx := context.TODO()
+
+	c, err := client.New(ctx, sb.Address())
+	require.NoError(t, err)
+	defer c.Close()
+
+	run1 := llb.Image("busybox:latest").Run(llb.Args([]string{"sleep", "5"})).Root()
+	d1, err := run1.Marshal(ctx)
+	require.NoError(t, err)
+	run2 := llb.Image("busybox:latest").Run(llb.Args([]string{"sleep", "6"})).Root()
+	d2, err := run2.Marshal(ctx)
+	require.NoError(t, err)
+
+	timeStart := time.Now()
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		_, err := c.Solve(egCtx, d1, client.SolveOpt{}, nil)
+		return err
+	})
+	eg.Go(func() error {
+		_, err := c.Solve(egCtx, d2, client.SolveOpt{}, nil)
+		return err
+	})
+	err = eg.Wait()
+	require.NoError(t, err)
+
+	elapsed := time.Since(timeStart)
+
+	maxParallelism := sb.Value("max-parallelism")
+	if maxParallelism == maxParallelismSingle {
+		require.Greater(t, elapsed, 10*time.Second, "parallelism not restricted")
```

Comment:
Maybe we could use cache mounts to make this test faster. Would also like to test the case where parallelism is allowed. Maybe both processes create a file and than wait for up to 5 sec for it to get deleted by the other process. For the single case, the deletion does not happen.

Reply:
I can't get this to work. Instead of deletion I touched files. So

* Process 1: `touch /shared/signal1 && i=0; while [ ! -f /shared/signal2 ] && [ $i -lt 5 ]; do i=$((i+1)); sleep 1; done`
* Process 2: `touch /shared/signal2 && i=0; while [ ! -f /shared/signal1 ] && [ $i -lt 5 ]; do i=$((i+1)); sleep 1; done`

And then measured the time it took. It's not working in the standard non-rootless, non-stargz sandbox. From debug prints I'm seeing that process 1 gets a snapshot of the shared cache which isn't modifiable by process 2, while process 2 does see both process 1's changes and its own.

Rootless version works fine though. Could this be an implementation detail of the cache on overlayfs vs fuse-overlayfs?

<details>
  <summary>Sample output</summary>

```
        sandbox.go:224: time="2021-05-06T14:20:23Z" level=debug msg="> creating ru3wquozqvrehxd4b0f7vn799 [/bin/sh -c set -x; exec 2>&1; touch /shared/signal1 && i=0; while [ ! -f /shared/signal2 ] && [ $i -lt 20 ]; do i=$((i+1)); ls -al /shared; sleep 1; done]"
        sandbox.go:224: + exec
        sandbox.go:224: time="2021-05-06T14:20:23Z" level=debug msg="> creating q844iyt8zon1oskinyyf5vx9w [/bin/sh -c set -x; exec 2>&1; touch /shared/signal2 && i=0; while [ ! -f /shared/signal1 ] && [ $i -lt 20 ]; do i=$((i+1)); ls -al /shared; sleep 1; done]"
        sandbox.go:224: + exec
        sandbox.go:224: time="2021-05-06T14:20:24Z" level=debug msg="session finished: <nil>"
        sandbox.go:224: time="2021-05-06T14:20:44Z" level=debug msg="session finished: <nil>"
        sandbox.go:224: > sending sigterm 2021-05-06 14:20:44.260139677 +0000 UTC m=+41.207529603
        sandbox.go:224: time="2021-05-06T14:20:44Z" level=info msg="stopping server"
        sandbox.go:224: buildkitd: context canceled
        sandbox.go:224: > stopped 2021-05-06 14:20:44.264363927 +0000 UTC m=+41.211753811 exit status 1 1
        sandbox.go:221: stdout: /usr/bin/buildkitd
        sandbox.go:224: + touch /shared/signal1
        sandbox.go:224: + i=0
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 0 -lt 20 ]
        sandbox.go:224: + i=1
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + touch /shared/signal2
        sandbox.go:224: + i=0
        sandbox.go:224: + '[' '!' -f /shared/signal1 ]
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 1 -lt 20 ]
        sandbox.go:224: + i=2
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 2 -lt 20 ]
        sandbox.go:224: + i=3
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 3 -lt 20 ]
        sandbox.go:224: + i=4
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 4 -lt 20 ]
        sandbox.go:224: + i=5
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 5 -lt 20 ]
        sandbox.go:224: + i=6
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 6 -lt 20 ]
        sandbox.go:224: + i=7
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 7 -lt 20 ]
        sandbox.go:224: + i=8
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 8 -lt 20 ]
        sandbox.go:224: + i=9
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 9 -lt 20 ]
        sandbox.go:224: + i=10
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 10 -lt 20 ]
        sandbox.go:224: + i=11
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 11 -lt 20 ]
        sandbox.go:224: + i=12
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 12 -lt 20 ]
        sandbox.go:224: + i=13
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 13 -lt 20 ]
        sandbox.go:224: + i=14
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 14 -lt 20 ]
        sandbox.go:224: + i=15
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 15 -lt 20 ]
        sandbox.go:224: + i=16
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 16 -lt 20 ]
        sandbox.go:224: + i=17
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 17 -lt 20 ]
        sandbox.go:224: + i=18
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 18 -lt 20 ]
        sandbox.go:224: + i=19
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 19 -lt 20 ]
        sandbox.go:224: + i=20
        sandbox.go:224: + ls -al /shared
        sandbox.go:224: ls: /shared/signal2: No such file or directory
        sandbox.go:224: total 8
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 .
        sandbox.go:224: drwxr-xr-x    1 root     root          4096 May  6 14:20 ..
        sandbox.go:224: -rw-r--r--    1 root     root             0 May  6 14:20 signal1
        sandbox.go:224: + sleep 1
        sandbox.go:224: + '[' '!' -f /shared/signal2 ]
        sandbox.go:224: + '[' 20 -lt 20 ]
    --- PASS: TestJobsIntegration/TestParallelism/worker=oci-snapshotter-stargz/max-parallelism=single (38.92s)
    --- PASS: TestJobsIntegration/TestParallelism/worker=oci/max-parallelism=single (40.64s)
```

</details>

Reply:
I think I figure it out.